test/unit/BPool/BPool.t.sol

```diff
@@ -50,4 +72,203 @@ contract BPool is BPoolBase {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.getBalance(tokens[0]);
+  }
+
+  function test_GetBalanceRevertWhen_TokenIsNotBound() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    bPool.getBalance(unknownToken);
+  }
+
+  function test_GetBalanceWhenPreconditionsAreMet(uint256 tokenBalance) external {
+    vm.mockCall(tokens[0], abi.encodePacked(IERC20.balanceOf.selector), abi.encode(tokenBalance));
+    // it queries token balance
+    vm.expectCall(tokens[0], abi.encodeWithSelector(IERC20.balanceOf.selector));
+    // it returns token balance
+    uint256 _balance = bPool.getBalance(tokens[0]);
+    assertEq(_balance, tokenBalance);
+  }
+
+  function test_GetSwapFeeRevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.getSwapFee();
+  }
+
+  function test_GetSwapFeeWhenPreconditionsAreMet() external view {
+    // it returns swap fee
+    uint256 _swapFee = bPool.getSwapFee();
+    assertEq(_swapFee, swapFee);
+  }
+
+  function test_GetControllerRevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.getController();
+  }
+
+  function test_GetControllerWhenPreconditionsAreMet() external view {
+    // it returns controller
+    address _controller = bPool.getController();
+    assertEq(_controller, controller);
+  }
+
+  function test_GetSpotPriceRevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.getSpotPrice(tokens[0], tokens[1]);
+  }
+
+  function test_GetSpotPriceRevertWhen_TokenInIsNotBound() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    bPool.getSpotPrice(unknownToken, tokens[1]);
+  }
+
+  function test_GetSpotPriceRevertWhen_TokenOutIsNotBound() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    bPool.getSpotPrice(tokens[0], unknownToken);
+  }
+
+  function test_GetSpotPriceWhenPreconditionsAreMet() external {
+    // it queries token in balance
+    vm.mockCall(tokens[0], abi.encodePacked(IERC20.balanceOf.selector), abi.encode(balanceTokenIn));
+    vm.expectCall(tokens[0], abi.encodeWithSelector(IERC20.balanceOf.selector));
+    // it queries token out balance
+    vm.mockCall(tokens[1], abi.encodePacked(IERC20.balanceOf.selector), abi.encode(balanceTokenOut));
+    vm.expectCall(tokens[1], abi.encodeWithSelector(IERC20.balanceOf.selector));
+    // it returns spot price
+    uint256 _spotPrice = calcSpotPrice(balanceTokenIn, tokenWeight, balanceTokenOut, tokenWeight, swapFee);
```

Comment:
I'd hard-code the value at the top in a 'valid scenario', to avoid testing the implementation against itself